asynciterator.ts

```diff
@@ -1852,17 +1889,167 @@ class HistoryReader<T> {
   }
 }
 
+/**
+ * The IteratorIterator class provides a performant wrapper
+ * to convert ES Iterable instances into AsyncIterator instances.
+ */
+export class IteratorIterator<T> extends AsyncIterator<T> {
+  /**
+   *
+   * @param source
+   */
+  constructor(private source: Iterator<T>) {
+    super();
+    this.readable = true;
+  }
+
+  read(): T | null {
+    const item = this.source.next();
+    if (item.done) {
+      this.close();
+      return null;
+    }
+    return item.value;
+  }
+}
+
+/**
+ * The WrappingIterator class can be used to wrap all supported kinds of data
+ * sources, both promisified and not, into AsyncIterator instances.
+ * This class is not meant to be used directly but, rather, through the
+ * `wrap()` function as it provides the mechanism that allows the latter
+ * to synchronously return an instance of AsyncIterator even for promisified
+ * sources.
+ */
+export class WrappingIterator<T> extends AsyncIterator<T> {
+  protected _source?: AsyncIteratorLike<T>;
+
+  constructor(sourceOrPromise: WrapSource<T> | Promise<WrapSource<T>> | null | undefined, options: WrapOptions = {}) {
+    super();
+    this._onSourceEnd = this._onSourceEnd.bind(this);
+    this._onSourceError = this._onSourceError.bind(this);
+    this._onSourceReadable = this._onSourceReadable.bind(this);
+    if (sourceOrPromise)
+      this.setSource(sourceOrPromise, options);
+  }
+
+  setSource(sourceOrPromise: WrapSource<T> | Promise<WrapSource<T>>, options: WrapOptions = {}) {
+    if (isPromise(sourceOrPromise)) {
+      sourceOrPromise
+        .then(source => this._wrapSource<T>(source, options))
+        .catch(err => this.emit('error', err));
+    }
+    else {
+      this._wrapSource<T>(sourceOrPromise, options);
+    }
+  }
+
+  protected _wrapSource<T>(source: WrapSource<T>, options: WrapOptions) {
+    try {
+      // eslint-disable-next-line no-use-before-define
+      const wrappedSource = isAsyncIteratorLike<T>(source) ? source : _wrap<T>(source, options);
+      wrappedSource.on('end', this._onSourceEnd);
+      wrappedSource.on('error', this._onSourceError);
+      wrappedSource.on('readable', this._onSourceReadable);
+      // @ts-ignore
+      this._source = wrappedSource;
+      this.readable = true;
+    }
+    catch (err) {
+      scheduleTask(() => this.emit('error', err));
+    }
+  }
+
+  protected _onSourceReadable() {
+    if (this.readable)
+      this.emit('readable');
+    else
+      this.readable = true;
+  }
+
+  protected _onSourceEnd() {
+    this.close();
+    this._cleanup();
+  }
+
```

Comment:
You can follow the same logic as [here](https://github.com/RubenVerborgh/AsyncIterator/blob/6b4f039d5d12dc48256695f4133c70a931a490b7/asynciterator.ts#L1255-L1264) and move these to being functions outside the class. Put the `_cleanup` logic in `close` and just [reflect the readable status of the source](https://github.com/RubenVerborgh/AsyncIterator/blob/6b4f039d5d12dc48256695f4133c70a931a490b7/asynciterator.ts#L1269-L1275)

Reply:
I can move those functions outside of the `WrappingIterator` class but I can't reflect the readable status of the source as I cannot assume the latter to be an instance of `AsyncIterator` and/or have a boolean `readable` property. I notice that `stream.Readable` has had that property since Node.js `v11.4.0` but this package supports back to `v10`.

Reply:
I also can't add the `_destination` property to `WrappingIterator#_source` as the latter can be user-provided (if we detect it to be an `AsyncIteratorLike`)... I'll add a couple of dedicated functions.

Reply:
I ended up moving cleanup logic into `protected _end()` but I can't think of other clean ways to improve on the above:

1) I can't reflect the readable status of the source because it might not have one (RDF/JS streams, `stream.Readable` when using Node.js version `[10.0 - 11.4), ...`.
2) I can't add `_destination` to `WrappingIterator#_source` or modify the source in any way as it might be user-provided, so I can't use `destinationEmitError` and the likes.
3) I still need to keep references to event listeners around so that I can call `removeListener()` when the wrapping iterator is closed / destroyed.

Leaving this open in case you have suggestions @jeswr , otherwise feel free to resolve.